go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,316 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	utilnet "k8s.io/utils/net"
+
+	egressfirewallapi "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressfirewall/v1"
+
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog"
+)
+
+const (
+	defaultStartPriority = 2000
+)
+
+type egressFirewall struct {
+	name        string
+	namespace   string
+	egressRules []*egressFirewallRule
+}
+
+type egressFirewallRule struct {
+	id     int
+	access egressfirewallapi.EgressFirewallRuleType
+	ports  []port
+	to     destination
+}
+
+type port struct {
+	protocol string
+	portNum  int32
+}
+
+type destination struct {
+	cidrSelector string
+}
+
+func newEgressFirewall(egressFirewallPolicy *egressfirewallapi.EgressFirewall) *egressFirewall {
+	ef := &egressFirewall{
+		name:        egressFirewallPolicy.Name,
+		namespace:   egressFirewallPolicy.Namespace,
+		egressRules: make([]*egressFirewallRule, 0),
+	}
+	return ef
+}
+
+func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewallRule, id int) (*egressFirewallRule, error) {
+	efr := &egressFirewallRule{
+		id:     id,
+		access: rawEgressFirewallRule.Type,
+	}
+
+	if rawEgressFirewallRule.To.CIDRSelector == "" {
+		return nil, fmt.Errorf("EgressFirewallRule must have a CIDRselector set")
+	}
+
+	_, _, err := net.ParseCIDR(rawEgressFirewallRule.To.CIDRSelector)
+	if err != nil {
+		return nil, err
+
+	}
+	efr.to.cidrSelector = rawEgressFirewallRule.To.CIDRSelector
+
+	for _, rawPort := range rawEgressFirewallRule.Ports {
+		port := port{}
+		if rawPort.Protocol != "" {
+			port.protocol = rawPort.Protocol
+		}
+		if rawPort.PortNum != 0 {
```

Comment:
Ditto here: https://github.com/ovn-org/ovn-kubernetes/blob/25861041266b38e64e44665c1742833652f85265/go-controller/pkg/crd/egressfirewall/v1/types.go#L66

But maybe you can test that the CRD validation actually does its job?

Reply:
CRD validates value